### PR TITLE
unify pytorch spark estimaor usage with ray backend

### PR DESCRIPTION
**Pull request by Le-Zheng**

### CODE REVIEW

File: docs/readthedocs/source/doc/Orca/QuickStart/orca-pytorch-quickstart.md [hunk 1]
```diff
@@ -70,10 +70,17 @@ class LeNet(nn.Module):
...
-adam = torch.optim.Adam(model.parameters(), 0.001)
+```
+After defining your model, you need to define a *Model Creator Function* that returns an instance of your model, and a *Optimizer Creator Function* that returns a PyTorch optimizer.
+
+```python
+def model_creator(config):
+    model = LeNet()
+    return model
+
+def optim_creator(model, config):
```
Comment (qiuxin2012) [inline docs/readthedocs/source/doc/Orca/QuickStart/orca-pytorch-quickstart.md:82]: what does this config save something? learning rate or something else?

Comment (Le-Zheng) [reply docs/readthedocs/source/doc/Orca/QuickStart/orca-pytorch-quickstart.md:82]: learning rate, updated.

**Comment by Le-Zheng** [comment]
jenkins passed: http://10.239.47.210:18888/job/ZOO-PR-Validation/5497/

File: docs/readthedocs/source/doc/Orca/Overview/distributed-training-inference.md [hunk 2]
```diff
@@ -106,18 +106,24 @@ View the related [Python API doc]() for more details.
 Users may create a PyTorch `Estimator` using the *BigDL* backend (currently default for PyTorch) as follows:
 
 ```python
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:108]: do not change this example

File: docs/docs/colab-notebook/orca/quickstart/pytorch_lenet_mnist.ipynb [hunk 3]
```diff
@@ -299,12 +299,28 @@
     "        x = self.fc2(x)\n",
```
Comment (jason-dai) [inline docs/docs/colab-notebook/orca/quickstart/pytorch_lenet_mnist.ipynb:299]: do not change this example

File: docs/readthedocs/source/doc/Orca/QuickStart/orca-pytorch-quickstart.md [hunk 4]
```diff
@@ -70,15 +70,22 @@ class LeNet(nn.Module):
         x = self.fc2(x)
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/Orca/QuickStart/orca-pytorch-quickstart.md:70]: do not change this example

**Comment by jason-dai** [comment]
How do we properly handle batch_size in this case? When the input data is, say, Spark Dataframe, we ask the user to pass total batch size; then we calculate the per-worker batch and apply batching automatically. But if the input is Dataloader creator function, the batch_size specified by the user is per-worker batch size? If the input is Dataloader , batch_size is ignored? @qiuxin2012

**Comment by qiuxin2012** [comment]
> How do we properly handle batch_size in this case? When the input data is, say, Spark Dataframe, we ask the user to pass total batch size; then we calculate the per-worker batch and apply batching automatically. But if the input is Dataloader creator function, the batch_size specified by the user is per-worker batch size? If the input is Dataloader , batch_size is ignored? @qiuxin2012

The batch_size is the same with Spark Dataframe, it's also the total batchsize, we calculate the per-worker batch and apply batching automatically.
See https://github.com/intel-analytics/analytics-zoo/blob/df747311f188633b27745013abf6ecc839b0ae24/pyzoo/zoo/feature/common.py#L381, we will check the batchsize and apply a new batchsize in createFeatureSetFromPyTorch.

**Comment by jason-dai** [comment]
> > How do we properly handle batch_size in this case? When the input data is, say, Spark Dataframe, we ask the user to pass total batch size; then we calculate the per-worker batch and apply batching automatically. But if the input is Dataloader creator function, the batch_size specified by the user is per-worker batch size? If the input is Dataloader , batch_size is ignored? @qiuxin2012
> 
> The batch_size is the same with Spark Dataframe, it's also the total batchsize, we calculate the per-worker batch and apply batching automatically.
> See
> 
> https://github.com/intel-analytics/analytics-zoo/blob/df747311f188633b27745013abf6ecc839b0ae24/pyzoo/zoo/feature/common.py#L381
> 
> , we will check the batchsize and apply a new batchsize in createFeatureSetFromPyTorch.

When the user passes a Dataloader as input in `fit`, the batch size is completely ignored?

**Comment by qiuxin2012** [comment]
> > > How do we properly handle batch_size in this case? When the input data is, say, Spark Dataframe, we ask the user to pass total batch size; then we calculate the per-worker batch and apply batching automatically. But if the input is Dataloader creator function, the batch_size specified by the user is per-worker batch size? If the input is Dataloader , batch_size is ignored? @qiuxin2012
> > 
> > 
> > The batch_size is the same with Spark Dataframe, it's also the total batchsize, we calculate the per-worker batch and apply batching automatically.
> > See
> > https://github.com/intel-analytics/analytics-zoo/blob/df747311f188633b27745013abf6ecc839b0ae24/pyzoo/zoo/feature/common.py#L381
> > 
> > , we will check the batchsize and apply a new batchsize in createFeatureSetFromPyTorch.
> 
> When the user passes a Dataloader as input in `fit`, the batch size is completely ignored?

Yes, it's ignored.
https://github.com/intel-analytics/analytics-zoo/blob/df747311f188633b27745013abf6ecc839b0ae24/pyzoo/zoo/orca/learn/pytorch/estimator.py#L376

**Comment by Le-Zheng** [comment]
> > > > How do we properly handle batch_size in this case? When the input data is, say, Spark Dataframe, we ask the user to pass total batch size; then we calculate the per-worker batch and apply batching automatically. But if the input is Dataloader creator function, the batch_size specified by the user is per-worker batch size? If the input is Dataloader , batch_size is ignored? @qiuxin2012
> > > 
> > > 
> > > The batch_size is the same with Spark Dataframe, it's also the total batchsize, we calculate the per-worker batch and apply batching automatically.
> > > See
> > > https://github.com/intel-analytics/analytics-zoo/blob/df747311f188633b27745013abf6ecc839b0ae24/pyzoo/zoo/feature/common.py#L381
> > > 
> > > , we will check the batchsize and apply a new batchsize in createFeatureSetFromPyTorch.
> > 
> > 
> > When the user passes a Dataloader as input in `fit`, the batch size is completely ignored?
> 
> Yes, it's ignored.
> https://github.com/intel-analytics/analytics-zoo/blob/df747311f188633b27745013abf6ecc839b0ae24/pyzoo/zoo/orca/learn/pytorch/estimator.py#L376

Shall we check if data loader batch size and `fit` batch size are consistent and remind a user to input the same batch size?

**Comment by jason-dai** [comment]
> > > > > > > How do we properly handle batch_size in this case? When the input data is, say, Spark Dataframe, we ask the user to pass total batch size; then we calculate the per-worker batch and apply batching automatically. But if the input is Dataloader creator function, the batch_size specified by the user is per-worker batch size? If the input is Dataloader , batch_size is ignored? @qiuxin2012
> > > > > > 
> > > > > > 
> > > > > > The batch_size is the same with Spark Dataframe, it's also the total batchsize, we calculate the per-worker batch and apply batching automatically.
> > > > > > See
> > > > > > https://github.com/intel-analytics/analytics-zoo/blob/df747311f188633b27745013abf6ecc839b0ae24/pyzoo/zoo/feature/common.py#L381
> > > > > > 
> > > > > > , we will check the batchsize and apply a new batchsize in createFeatureSetFromPyTorch.
> > > > > 
> > > > > 
> > > > > When the user passes a Dataloader as input in `fit`, the batch size is completely ignored?
> > > > 
> > > > 
> > > > Yes, it's ignored.
> > > > https://github.com/intel-analytics/analytics-zoo/blob/df747311f188633b27745013abf6ecc839b0ae24/pyzoo/zoo/orca/learn/pytorch/estimator.py#L376
> > > 
> > > 
> > > Shall we check if data loader batch size and `fit` batch size are consistent and remind a user to input the same batch size?
> > 
> > 
> > We should make sure that when a data loader is passed in, the data loader is already batched and batch size is None; for the other cases, the user need to specify batch size.
> 
> I think we can add a warning: If batch size is not None, we sould notice our user the batchsize is ignored as batch size is already defined in dataloader.

User may simply ignore the warning; we should enforce that (1) the dataloader is batched, and (2) the user should not pass in batch size in this case.

**Comment by Le-Zheng** [comment]
> > > > > > > > How do we properly handle batch_size in this case? When the input data is, say, Spark Dataframe, we ask the user to pass total batch size; then we calculate the per-worker batch and apply batching automatically. But if the input is Dataloader creator function, the batch_size specified by the user is per-worker batch size? If the input is Dataloader , batch_size is ignored? @qiuxin2012
> > > > > > > 
> > > > > > > 
> > > > > > > The batch_size is the same with Spark Dataframe, it's also the total batchsize, we calculate the per-worker batch and apply batching automatically.
> > > > > > > See
> > > > > > > https://github.com/intel-analytics/analytics-zoo/blob/df747311f188633b27745013abf6ecc839b0ae24/pyzoo/zoo/feature/common.py#L381
> > > > > > > 
> > > > > > > , we will check the batchsize and apply a new batchsize in createFeatureSetFromPyTorch.
> > > > > > 
> > > > > > 
> > > > > > When the user passes a Dataloader as input in `fit`, the batch size is completely ignored?
> > > > > 
> > > > > 
> > > > > Yes, it's ignored.
> > > > > https://github.com/intel-analytics/analytics-zoo/blob/df747311f188633b27745013abf6ecc839b0ae24/pyzoo/zoo/orca/learn/pytorch/estimator.py#L376
> > > > 
> > > > 
> > > > Shall we check if data loader batch size and `fit` batch size are consistent and remind a user to input the same batch size?
> > > 
> > > 
> > > We should make sure that when a data loader is passed in, the data loader is already batched and batch size is None; for the other cases, the user need to specify batch size.
> > 
> > 
> > I think we can add a warning: If batch size is not None, we sould notice our user the batchsize is ignored as batch size is already defined in dataloader.
> 
> User may simply ignore the warning; we should enforce that (1) the dataloader is batched, and (2) the user should not pass in batch size in this case.

For (1) we may check dataloader.batch_size is not None. For (2) the `fit` method has default batch_size=32. Shall we change the default batch_size=None and tell user not to specify batchsize in 'fit' when passing dataloader? For the other cases, the user must specify batch size.

**Comment by jason-dai** [comment]
> > > > > > > > > How do we properly handle batch_size in this case? When the input data is, say, Spark Dataframe, we ask the user to pass total batch size; then we calculate the per-worker batch and apply batching automatically. But if the input is Dataloader creator function, the batch_size specified by the user is per-worker batch size? If the input is Dataloader , batch_size is ignored? @qiuxin2012
> > > > > > > > 
> > > > > > > > 
> > > > > > > > The batch_size is the same with Spark Dataframe, it's also the total batchsize, we calculate the per-worker batch and apply batching automatically.
> > > > > > > > See
> > > > > > > > https://github.com/intel-analytics/analytics-zoo/blob/df747311f188633b27745013abf6ecc839b0ae24/pyzoo/zoo/feature/common.py#L381
> > > > > > > > 
> > > > > > > > , we will check the batchsize and apply a new batchsize in createFeatureSetFromPyTorch.
> > > > > > > 
> > > > > > > 
> > > > > > > When the user passes a Dataloader as input in `fit`, the batch size is completely ignored?
> > > > > > 
> > > > > > 
> > > > > > Yes, it's ignored.
> > > > > > https://github.com/intel-analytics/analytics-zoo/blob/df747311f188633b27745013abf6ecc839b0ae24/pyzoo/zoo/orca/learn/pytorch/estimator.py#L376
> > > > > 
> > > > > 
> > > > > Shall we check if data loader batch size and `fit` batch size are consistent and remind a user to input the same batch size?
> > > > 
> > > > 
> > > > We should make sure that when a data loader is passed in, the data loader is already batched and batch size is None; for the other cases, the user need to specify batch size.
> > > 
> > > 
> > > I think we can add a warning: If batch size is not None, we sould notice our user the batchsize is ignored as batch size is already defined in dataloader.
> > 
> > 
> > User may simply ignore the warning; we should enforce that (1) the dataloader is batched, and (2) the user should not pass in batch size in this case.
> 
> For (1) we may check dataloader.batch_size is not None. For (2) the `fit` method has default batch_size=32. Shall we change the default batch_size=None and tell user not to specify batchsize in 'fit' when passing dataloader? For the other cases, the user must specify batch size.

Yes, default to None and add checking code in `fit`

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 5]
```diff
@@ -359,7 +359,11 @@ def fit(self, data, epochs=1, batch_size=32, feature_cols=None, label_cols=None,
         from zoo.orca.learn.trigger import Trigger
 
         end_trigger = MaxEpoch(epochs)
-        assert batch_size > 0, "batch_size should be greater than 0"
+        if isinstance(data, DataLoader):
+            assert batch_size is None and data.batch_size > 0, "You should specify batch size in" \
+            "DataLoader."
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:364]: You should explain that the user should not pass `batch_size` to `fit` and should set batch size in PyTorch DataLoader.

Comment (Le-Zheng) [reply pyzoo/zoo/orca/learn/pytorch/estimator.py]: Ok. Updated as "You should specify batch size in PyTorch DataLoader, not in fit method. PyTorch DataLoader supports automatic batching(default)."

**Comment by Le-Zheng** [comment]
jenkins passed: http://10.239.176.111:18888/view/ZOO-PR/job/ZOO-PR-Validation/5562/

File: pyzoo/zoo/orca/learn/pytorch/estimator.py (hunk 5 above)
Comment (jason-dai) [reply pyzoo/zoo/orca/learn/pytorch/estimator.py]: "When using PyTorch Dataloader as input, you need to specify the batch size in DataLoader and don't specify batch_size in the fit method"

Comment (Le-Zheng) [reply pyzoo/zoo/orca/learn/pytorch/estimator.py]: Updated
